### PR TITLE
Add tests for issue #3736

### LINKED ISSUE #3736: MethodError with empty summations
Reported on discourse: https://discourse.julialang.org/t/weird-empty-sum-behavior-in-jump/113386

Not sure if we should fix this in JuMP or MA.

```julia
julia> using JuMP

julia> model = Model();

julia> @variable(model, x);

julia> @expression(model, -x^2 * sum(x for i in 1:0))
ERROR: MethodError: no method matching zero(::Type{MutableArithmetics.Zero})

Closest candidates are:
  zero(::Type{Union{}}, Any...)
   @ Base number.jl:310
  zero(::Type{Pkg.Resolve.FieldValue})
   @ Pkg ~/.julia/juliaup/julia-1.10.2+0.x64.apple.darwin14/share/julia/stdlib/v1.10/Pkg/src/Resolve/fieldvalues.jl:38
  zero(::Type{Missing})
   @ Base missing.jl:104
  ...

Stacktrace:
  [1] _instantiate_zero(::Type{MutableArithmetics.Zero})
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:28
  [2] promote_operation_fallback(op::typeof(*), ::Type{QuadExpr}, ::Type{MutableArithmetics.Zero})
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:51
  [3] promote_operation(::typeof(*), ::Type, ::Type)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:113
  [4] mutability(::Type, ::Function, ::Type, ::Type)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:273
  [5] mutability(::QuadExpr, ::Function, ::QuadExpr, ::MutableArithmetics.Zero)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:281
  [6] operate!!(::typeof(*), ::QuadExpr, ::MutableArithmetics.Zero)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:616
  [7] macro expansion
    @ ~/.julia/dev/MutableArithmetics/src/rewrite.jl:321 [inlined]
  [8] macro expansion
    @ ~/.julia/dev/JuMP/src/macros.jl:257 [inlined]
  [9] macro expansion
    @ ~/.julia/dev/JuMP/src/macros/@expression.jl:86 [inlined]
 [10] macro expansion
    @ ~/.julia/dev/JuMP/src/macros.jl:393 [inlined]
 [11] top-level scope
    @ ./REPL[37]:1
```

### PR DESCRIPTION
**Pull request by odow**
Closes #3736

Needs release with https://github.com/jump-dev/MutableArithmetics.jl/pull/283 before merging.

```Julia
julia> function test_multiply_expr_MA_Zero()
           model = Model()
           @variable(model, x)
           for f in (x, x^2, sin(x))
               @test @expression(model, f * sum(x for i in 1:0)) == 0.0
               @test @expression(model, sum(x for i in 1:0) * f) == 0.0
               @test @expression(model, -f * sum(x for i in 1:0)) == 0.0
               @test @expression(model, sum(x for i in 1:0) * -f) == 0.0
               @test @expression(model, (f + f) * sum(x for i in 1:0)) == 0.0
               @test @expression(model, sum(x for i in 1:0) * (f + f)) == 0.0

               @test isequal_canonical(@expression(model, f + sum(x for i in 1:0)), f)
               @test isequal_canonical(@expression(model, sum(x for i in 1:0) + f), f)
               @test isequal_canonical(
                   @expression(model, -f + sum(x for i in 1:0)),
                   -1.0 * f,  # Needed for f = sin(x)
               )
               @test isequal_canonical(
                   @expression(model, sum(x for i in 1:0) + -f),
                   -1.0 * f,  # Needed for f = sin(x)
               )
               @test isequal_canonical(
                   @expression(model, (f + f) + sum(x for i in 1:0)),
                   f + f,
               )
               @test isequal_canonical(
                   @expression(model, sum(x for i in 1:0) + (f + f)),
                   f + f,
               )
           end
           return
       end
test_multiply_expr_MA_Zero (generic function with 1 method)

julia> @testset "" begin
           test_multiply_expr_MA_Zero()
       end
Test Summary: | Pass  Total  Time
              |   36     36  0.0s
Test.DefaultTestSet("", Any[], 36, false, false, true, 1.713934627962967e9, 1.713934627963343e9, false, "REPL[60]")
```

### CODE REVIEW

File: src/operators.jl [hunk 1]
```diff
@@ -568,3 +568,9 @@ function Base.complex(
 )
     return r + im * i
 end
+
+# These methods are used to provide an efficient implementation for a common
+# case like `x^2 * sum(f for i in 1:0)`, which lowers to
+# `_MA.operate!!(*, x^2, _MA.Zero())`
+_MA.operate!!(::typeof(*), ::AbstractJuMPScalar, ::_MA.Zero) = _MA.Zero()
```
Comment (blegat) [inline src/operators.jl:575]: There are already methods for that in MA, is that to resolve an ambiguity ?

**Comment by odow** [comment]
Where are the methods?

I added because the stack trace from #3736 includes
```julia
  [1] _instantiate_zero(::Type{MutableArithmetics.Zero})
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:28
  [2] promote_operation_fallback(op::typeof(*), ::Type{QuadExpr}, ::Type{MutableArithmetics.Zero})
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:51
  [3] promote_operation(::typeof(*), ::Type, ::Type)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:113
  [4] mutability(::Type, ::Function, ::Type, ::Type)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:273
  [5] mutability(::QuadExpr, ::Function, ::QuadExpr, ::MutableArithmetics.Zero)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:281
  [6] operate!!(::typeof(*), ::QuadExpr, ::MutableArithmetics.Zero)
    @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:616
```
This is going to additionally call `zero(QuadExpr)` to check whether the expression is mutable.

**Comment by blegat** [comment]
Here:
https://github.com/jump-dev/MutableArithmetics.jl/blob/a6ed0f5ef5dfe67e463382d0b2d9ada60d77c673/src/rewrite.jl#L61-L62

**Comment by odow** [comment]
Changed to implement `promote_operation`

**Comment by blegat** [comment]
This should be in MA next to
https://github.com/jump-dev/MutableArithmetics.jl/blob/a6ed0f5ef5dfe67e463382d0b2d9ada60d77c673/src/rewrite.jl#L61

**Comment by odow** [comment]
See https://github.com/jump-dev/MutableArithmetics.jl/pull/284

**Pull request merged**